src/core/thread/mesh_forwarder_ftd.cpp

```diff
@@ -676,6 +678,14 @@ void MeshForwarder::HandleMesh(FrameData &aFrameData, const Mac::Address &aMacSo
     meshAddrs.mDestination.SetShort(meshHeader.GetDestination());
 
     UpdateRoutes(aFrameData, meshAddrs);
+    macSource  = Get<RouterTable>().FindRouterByRloc16(aMacSource.GetShort());
+    meshSource = Get<RouterTable>().FindRouterByRloc16(meshAddrs.mSource.GetShort());
+    if (macSource != nullptr && meshSource != nullptr && !Get<RouterTable>().FindNextHopOf(*meshSource))
+    {
+        Get<RouterTable>().SetRouterNextHopAndCost(*meshSource, macSource->GetRouterId(),
+                                                   Mle::kMaxRouteCost - 1 -
+                                                       Get<RouterTable>().GetPathCost(aMacSource.GetShort()));
+    }
```

Comment:
How about adding check `macSource != meshSource` to the `if`?
Basically not updating on the very first hop? In that case there should be already a link between the two nodes.

Reply:
Another scenario to consider is when the originator of the mesh header message is an FTD child. On its parent node, both macSource and meshSource would reference the parent, but we wouldn't want to add a next hop in this case as it would point back to the parent itself.